### PR TITLE
Implement RRT and RRT*

### PR DESCRIPTION
**Pull request by pranavjad**

### CODE REVIEW

File: .vscode/c_cpp_properties.json
Comment (zkingston) [inline .vscode/c_cpp_properties.json:1]: Remove this file and put it in the .gitignore

File: .vscode/settings.json
Comment (zkingston) [inline .vscode/settings.json:1]: Remove this file and put it in the .gitignore

File: .vscode/tasks.json
Comment (zkingston) [inline .vscode/tasks.json:1]: Same, remove and ignore

File: scripts/ao_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+pd.set_option('display.max_columns', None)
+
+# Starting configuration
+start = [0., -0.785, 0., -2.356, 0., 1.571, 0.785]
+
+# Goal configuration
+goal = [2.35, 1., 0., -0.8, 0, 2.5, 0.785]
+
+print(np.sqrt(sum([(start[i] - goal[i])**2 for i in range(len(start))])))
```
Comment (zkingston) [inline scripts/ao_benchmark.py:18]: Remove

File: scripts/ao_benchmark.py
Comment (zkingston) [inline scripts/ao_benchmark.py:1]: Make sure to run yapf formatting

File: scripts/ao_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    all_results = []
+    spheres = [np.array(sphere) for sphere in problem]
+    max_iters = np.arange(1, 100000, 1000)
+
+    env = vamp.Environment()
+    for sphere in spheres:
+        env.add_sphere(vamp.Sphere(sphere, radius))
+    
+    plan_settings.force_max_iters = True
+    plan_settings.range = 3.0
```
Comment (zkingston) [inline scripts/ao_benchmark.py:60]: Make a CL parameter

File: scripts/flying_sphere.py [hunk 3]
```diff
@@ -49,8 +50,8 @@ def main(
 
         pb.set_camera([x + 1, y + 1, z + 1], [0, 0, 0])
 
-        while True:
-            ...
+        while pybullet.isConnected():
+            pybullet.stepSimulation()
```
Comment (zkingston) [inline scripts/flying_sphere.py:55]: Can you move to the PyBulletSimulator class implementation and add a function "wait()" or something equivalent?

File: scripts/sphere_cage_example.py [hunk 4]
```diff
@@ -12,6 +12,9 @@
 # Goal configuration
 b = [2.35, 1., 0., -0.8, 0, 2.5, 0.785]
 
+print("optimal cost: ", np.sqrt(sum([(a[i] - b[i])**2 for i in range(len(a))])))
```
Comment (zkingston) [inline scripts/sphere_cage_example.py:15]: Remove

File: scripts/sphere_cage_example.py [hunk 5]
```diff
@@ -43,7 +46,12 @@ def main(
 
     (vamp_module, planner_func, plan_settings,
      simp_settings) = vamp.configure_robot_and_planner_with_kwargs("panda", planner, **kwargs)
-
+    
+    # rrt* settings
+    # plan_settings.max_iterations = 100000
+    # plan_settings.force_max_iters = True
+    # plan_settings.range = 5.0
+    # plan_settings.rewire_factor = 3.0
```
Comment (zkingston) [inline scripts/sphere_cage_example.py:54]: Remove

Comment (zkingston) [reply scripts/sphere_cage_example.py:54]: Should be able to set these via command line, please check if that works.

File: src/impl/vamp/planning/rrt.hh
Comment (zkingston) [inline src/impl/vamp/planning/rrt.hh:1]: For the actual PR I think we remove vanilla RRT, I don't want to clutter up the VAMP codebase with every SBMP like OMPL

File: src/impl/vamp/planning/rrt_star.hh [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+            const auto buffer_index = [&buffer](std::size_t index) -> float *
+            {
+                return buffer.get() + index * Configuration::num_scalars_rounded;
+            };
+
+            std::vector<std::size_t> parent(settings.max_samples);
+            std::vector<std::vector<std::size_t>> children(settings.max_samples, std::vector<std::size_t>());
+            std::vector<float> cost(settings.max_samples, -1.0);
+
+            const auto add_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node) -> void
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:60]: No need to specify -> void at the end

File: src/impl/vamp/planning/rrt_star.hh [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+            {
+                parent[child_node] = parent_node;
+                children[parent_node].push_back(child_node);
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:63]: Use emplace_back

File: src/impl/vamp/planning/rrt_star.hh
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:1]: Run clang-format

File: src/impl/vamp/planning/rrt_star.hh [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+            };
+            const auto remove_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node) -> void
+            {
+                parent[child_node] = -1;
+                auto it = std::find(children[parent_node].begin(), children[parent_node].end(), child_node);
+                if (it != children[parent_node].end()) children[parent_node].erase(it);
+            };
+            std::function<void(std::size_t)> update_children = [&cost, &children, &buffer_index, &update_children](std::size_t idx)
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:71]: const auto?

File: src/impl/vamp/planning/rrt_star.hh [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                {
+                    // insert new config into vertex set
+                    float *new_configuration_ptr = buffer_index(free_index);
+                    auto new_configuration = nearest_configuration + extension_vector;
+                    new_configuration.to_array(new_configuration_ptr);
+                    NNNode<dimension> new_node({free_index, {new_configuration_ptr}});
+
+                    // sample points in graph within radius r of new config
+                    float free_volume = space_measure();
+                    std::size_t card = tree.size() + 1; //plus 1 to account for the new configuration not inserted yet
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:147]: const

Comment (zkingston) [reply src/impl/vamp/planning/rrt_star.hh:147]: to all these values, actually

File: src/impl/vamp/planning/rrt_star.hh [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+                ))
+                {
+                    // insert new config into vertex set
+                    float *new_configuration_ptr = buffer_index(free_index);
+                    auto new_configuration = nearest_configuration + extension_vector;
+                    new_configuration.to_array(new_configuration_ptr);
+                    NNNode<dimension> new_node({free_index, {new_configuration_ptr}});
+
+                    // sample points in graph within radius r of new config
+                    float free_volume = space_measure();
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:146]: make static const at top, no need to compute each loop

Comment (zkingston) [reply src/impl/vamp/planning/rrt_star.hh:146]: for the constant variables, anyway

File: src/impl/vamp/planning/rrt_star.hh [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+                    // loop through near neighbors, find the one that gives us min cost from root
+                    std::vector<bool> collision_free(near.size());
+                    std::size_t idx = 0;
+                    for (auto &[node, distance] : near) {
+                        auto configuration = node.as_vector();
+                        collision_free[idx] = validate_motion<Robot, rake, resolution>(
+                            configuration, new_configuration, environment
+                        );
+                        float cur_cost = cost[node.index] + distance;
+                        if ((cur_cost < min_cost) && collision_free[idx]) {
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:170]: use `and` rather than `&&`

Comment (zkingston) [reply src/impl/vamp/planning/rrt_star.hh:170]: Move collision_free test first for early exit

File: src/impl/vamp/planning/rrt_star.hh [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+                    }
+                    if (!settings.force_max_iters && goal_reached) break;
+                    if (settings.force_max_iters && check_new_best) {
+                        // loop through goal nodes to find the best solution
+                        for (auto &[end_node_idx, goal]: goal_motions) {
+                            float cur_cost = cost[end_node_idx] + goal.distance(buffer_index(end_node_idx));
+                            if (cur_cost < best_path_cost) {
+                                best_path_cost = cur_cost;
+                                best_goal_motion = {end_node_idx, goal};
+                                // std::cout << "New best path found with cost: " << best_path_cost << " at iter " << iter << std::endl;
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:235]: remove

File: src/impl/vamp/planning/rrt_star.hh [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+            result.nanoseconds = vamp::utils::get_elapsed_nanoseconds(start_time);
+            result.iterations = iter;
+            result.size.emplace_back(tree.size());
+            return result;
+        }
+
+        inline static auto unitNballMeasure(unsigned int N) -> float
+        {
+            return std::pow(std::sqrt(M_PI), N) / std::tgamma(N / 2.0 + 1.0);
+        }
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:271]: constexpr?

File: src/impl/vamp/planning/rrt_star.hh [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+            tree.insert(NNNode<dimension>{start_index, {buffer_index(start_index)}});
+            parent[start_index] = start_index;
+            cost[start_index] = 0.0;
+
+            std::vector<std::pair<std::size_t, Configuration>> goal_motions;
+            float best_path_cost = std::numeric_limits<float>::max();
+            std::pair<std::size_t, Configuration> best_goal_motion;
+            // main loop
+            std::size_t iter = 0;
+            float unit_ball_volume = unitNballMeasure(dimension);
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:109]: constexpr? move to static?

File: src/impl/vamp/planning/rrt_star.hh [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+                                goal_reached = true;
+                                break;
+                            }
+                            else {
+                                goal_motions.push_back({free_index - 1, goal});
+                                check_new_best = true;
+                            }
+                        }
+                    }
+                    if (!settings.force_max_iters && goal_reached) break;
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:227]: `not` and `and`

File: src/vamp/pybullet_interface.py [hunk 16]
```diff
@@ -307,12 +307,12 @@ def add_environment_from_problem_dict(self, problem: Dict[str, Any], add_names:
                     )
 
     def draw_roadmap(self, fk_function, roadmap):
-        with DisableRendering(self.client), RedirectStream(sys.stdout), RedirectStream(sys.stderr):
-            for i, edge_list in enumerate(roadmap.edges):
-                for edge in edge_list:
-                    a = fk_function(roadmap[i].to_list())[-1]
-                    b = fk_function(roadmap[edge].to_list())[-1]
-                    self.client.addUserDebugLine([a.x, a.y, a.z], [b.x, b.y, b.z])
+        # with DisableRendering(self.client), RedirectStream(sys.stdout), RedirectStream(sys.stderr):
```
Comment (zkingston) [inline src/vamp/pybullet_interface.py:310]: Add back in - keep it the same for now but we need to figure out the Mac rendering issues later.

**Review by zkingston — CHANGES_REQUESTED** [review]
Some initial comments, make sure formatting is good and that you are following the general coding style of the rest of the project. I'd also just remove RRT and keep only RRT* for the actual contribution.

**Comment by pranavjad** [comment]
Created a new PR since this branch as a whole bunch of other stuff.

**Pull request closed without merging**